Add M3 blog article (#10806)
lampepfl/dotty: docs/blog/_posts/2020-12-18-scala3-m3.md

---
layout: blog-page
title: Scala 3.0.0-M3: developer's preview before RC1
author: Anatolii Kmetiuk
authorImg: /images/anatolii.png
date: 2020-12-18
---
We are happy to announce the release of Scala 3.0.0-M3. This release is the Developer's Preview release intended to contain all the features meant for RC1 which is tentatively planned for January 2021. The purpose of M3 is to give the community a chance to try out all the features and give us feedback before sealing them in RC1.

> To collect the feedback, we have designed a [Scala 3 Developer’s Preview satisfaction survey (3-5 min)](https://docs.google.com/forms/d/e/1FAIpQLSflVmTu9lhrtnSTh2tKAjUGrt3WvEgwlDqZg66O3EVSXd1aJg/viewform?usp=sf_link) – please fill it in to express how you feel about Scala 3.

Author: /cc @darjutak

For the weeks between M2 and M3, we were tying up all the loose ends so that to end 2020 having done everything intended for the release. Also, we have used this time as the last chance to review and fine-tune the language syntax. A large number of consultations with the community took place during this time so that to ensure everyone is on the same page.

In this article, you will find the summary of the most important changes that took place.

> 

Reviewer: Please, once this blog is merged: https://github.com/scala/scala-lang/pull/1195 add something like:

To read more about all things that surround the Scala 3 release in the next months, check out the "Scala 3 - Crossing the finish line" blog post on scala-lang.org

<!--more-->

> # SBT plugin bump

Author: /cc @smarter

Reviewer 2: ```suggestion
# sbt plugin update
```

SBT plugin was bumped to 0.5.0. Because of the changes in PR [#10607](https://github.com/lampepfl/dotty/pull/10607) this release of Scala 3 will not work with earlier versions of the SBT plugin. So if you are an early adopter of Scala 3 who has some projects in it, you will need to bump your SBT plugin version to migrate to M3.

> # Final syntactic tweaks

Author: /cc @odersky

given Ordering[Int]:
  ...
given [T: Ordering] as Ordering[List[T]]:
  ...

given global as ExecutionContext = ForkJoinContext()
given Context = ctx
```

Here is how the above is expressed now:

```scala
given intOrd: Ordering[Int] with
  ...
given listOrd[T: Ordering]: Ordering[List[T]] with
  ...

given Ordering[Int] with
  ...
given [T: Ordering]: Ordering[List[T]] with
  ...

given global: ExecutionContext = ForkJoinContext()
given Context = ctx
```

You can find an extensive discussion on the above change in its [PR #10538](https://github.com/lampepfl/dotty/pull/10538).

## Drop `as` in patterns
Since we dropped `as` from `given`s, we lost a strong reason for having `as` at all. For this reason, we dropped them also from patterns. If previously the correct binding syntax is patterns was:

```scala
case opt as Some(foo)
```

Now it got reverted to the Scala 2 style:

```scala
case opt @ Some(foo)
```

## Switch back to the old context function closure syntax
Previously, context function closures were written as follows:

```scala
(using s: Show[String]) => s.show("foobar")
```

Now, we have reverted to the following syntax:
```scala
(s: Show[String]) ?=> s.show("foobar")
```

> # `Matchable` trait – a new top-level type

Author: /cc @odersky

The motivation for this change is the fact that it is possible to access the underlying type referenced by an opaque type using pattern matching:

```scala
val imm = IArray(1,2,3) // supposedly immutable...
imm match
   case a: Array[Int] => a(0) = 0  // but that's shown to be lie
```

To address this change, a new trait, `Matchable`, is introduced at the top of the type hierarchy:

```scala
abstract class Any:
  def asInstanceOf
  def ==
  def !=
  def ##
  def equals
  def hashCode
  def toString

trait Matchable extends Any:

>   def isInstanceOf

Reviewer 3: Currently the documentation for `Matchable` claims that it has no methods by now but they will be moved to it in the future

  def getClass

class AnyVal extends Any, Matchable

class Object extends Any, Matchable
```

With the addition of this type, we now make the compiler emit warnings when trying to do a trick like the one discussed above:

```scala
-- Warning: i7314.scala:6:12 ---------------------------------------------------
6 |    case a: Array[Int] =>
  |            ^^^^^^^^^^
  |            pattern selector should be an instance of Matchable,
  |            but it has unmatchable type opaques.IArray[Int] instead
```

The change involved an extensive discussion which you can read in the [PR #10670](https://github.com/lampepfl/dotty/pull/10670).

# Tooling improvements
As the work on Scala 3 gets closer to the stable release, the focus increasingly shifts on the tooling available to get started with Scala 3.

> As part of the tooling effort, Scala 3 documentation tool is rapidly improved. [PR #10522](https://github.com/lampepfl/dotty/pull/10522) proves that the doctool can generate documentation for the community build projects.

Reviewer 4: I think we should mention that we moved away from dottydoc and we are using scala3doc by default now.

Reviewer 5: Seconding that. Judging by community build projects disabling the `doc` task, ppl had bad experience with Dottydoc previously, so it'd be nice to mention that the default has changed and that they can give generating documentation another shot.

[PR #10491](https://github.com/lampepfl/dotty/pull/10491) introduced the scripting capability to Scala 3. Consider the following source named e.g. `Main.scala`:

```scala
@main def Test(name: String): Unit =
  println(s"Hello ${name}!")
```

> If you have Scala 3 binaries on your path (which you can get by following the steps at the [Dotty website](https://dotty.epfl.ch/)'s section "Try DOtty"), you can run the `scala Main.scala World` command. This command will compile the source in question to a temporary directory and run the discovered main method with the argument `World`.

Reviewer 6: DOtty -> Dotty

Note the difference from the Scala 2 scripting implementation. In Scala 2, we do not require the users to have a `main` method in their scripts due to it being too cumbersome to write. In Scala 3, thanks to the top-level definitions and the `@main` annotations, `main` methods are one-liners and hence are more suited for scripts.

The documentation for this feature is available [here](https://dotty.epfl.ch/docs/usage/getting-started.html#scala-3-for-scripting).

> # Metaprogramming changes

Author: /cc @nicolasstucki

Reviewer 7: @anatoliykmetyuk the markdown format is wrong in this section `(..)[..]` vs `[..](..)`

Metaprogramming-wise, the majority of work is focused on polishing the API and making it more uniform. The following notable metaprogramming API changes occurred between M2 and M3:

> - Add reflect `MatchCase` `TypeRepr` (#10735)[https://github.com/lampepfl/dotty/pull/10735]

Reviewer 7: ```suggestion
- Add `Expr.asTerm` [#10694](https://github.com/lampepfl/dotty/pull/10694)
- Add reflect `MatchCase` `TypeRepr` (#10735)[https://github.com/lampepfl/dotty/pull/10735]
```

- Rework reflect Symbol fields API (#10705)[https://github.com/lampepfl/dotty/pull/10705]
- Remove `Expr.StringContext.unapply` (#10675)[https://github.com/lampepfl/dotty/pull/10675]
- Rename `Liftable` to `ToExpr` and `Unliftable` to `FromExpr` (#10618)[https://github.com/lampepfl/dotty/pull/10618]
- Remove Unliftable[Unit]  (#10570)[https://github.com/lampepfl/dotty/pull/10570]
- Remove reflect.LambdaType (#10548)[https://github.com/lampepfl/dotty/pull/10548]
- Add `scala.quoted.Expr.unapply` as dual of `Expr.apply` (#10580)[https://github.com/lampepfl/dotty/pull/10580]

> - Move `Quotes` as last parameter in `ExprMap.transform` (#10519)[https://github.com/lampepfl/dotty/pull/10519]

Reviewer 7: ```suggestion
- Move `Quotes` as last parameter in `ExprMap.transform` (#10519)[https://github.com/lampepfl/dotty/pull/10519]
- Rework reflect Constant API [#10753](https://github.com/lampepfl/dotty/pull/10753)
```

- Unify quoted.report and reflect.Reporting (#10474)[https://github.com/lampepfl/dotty/pull/10474]
- Fix #10359: Add GivenSelector to reflection API (#10469)[https://github.com/lampepfl/dotty/pull/10469]
- Rework reflect show API (#10661)[https://github.com/lampepfl/dotty/pull/10661]

> - Fix #6542: Pickle line sizes in TASTy (#10363)[https://github.com/lampepfl/dotty/pull/10363]

Reviewer 7: ```suggestion
```

---
layout: blog-page
title: Scala 3.0.0-M3: developer's preview before RC1
author: Anatolii Kmetiuk
authorImg: /images/anatolii.png
date: 2020-12-18
---
We are happy to announce the release of Scala 3.0.0-M3. This release is the Developer's Preview release intended to contain all the features meant for RC1 which is tentatively planned for January 2021. The purpose of M3 is to give the community a chance to try out all the features and give us feedback before sealing them in RC1.

To collect the feedback, we have designed a [Scala 3 Developer’s Preview satisfaction survey (3-5 min)](https://docs.google.com/forms/d/e/1FAIpQLSflVmTu9lhrtnSTh2tKAjUGrt3WvEgwlDqZg66O3EVSXd1aJg/viewform?usp=sf_link) – please fill it in to express how you feel about Scala 3.

For the weeks between M2 and M3, we were tying up all the loose ends so that to end 2020 having done everything intended for the release. Also, we have used this time as the last chance to review and fine-tune the language syntax. A large number of consultations with the community took place during this time so that to ensure everyone is on the same page.

> In this article, you will find the summary of the most important changes that took place.

Reviewer 2: ```suggestion
In this article, you will find the most important changes of this release compared to Scala 3.0.0-M2.
```

<!--more-->
# SBT plugin bump

> SBT plugin was bumped to 0.5.0. Because of the changes in PR [#10607](https://github.com/lampepfl/dotty/pull/10607) this release of Scala 3 will not work with earlier versions of the SBT plugin. So if you are an early adopter of Scala 3 who has some projects in it, you will need to bump your SBT plugin version to migrate to M3.

Reviewer 2: ```suggestion
We published a new version of the sbt plugin `sbt-dotty`, v0.5.0. Because of the changes in PR [#10607](https://github.com/lampepfl/dotty/pull/10607), this release of Scala 3 will not work with earlier versions of sbt-dotty. You will need to upgrade sbt-dotty to 0.5.0 to be able to use Scala 3.0.0-M3.
```

# Final syntactic tweaks
## `as` dropped from the `given` syntax
The following syntax is obsolete:

```scala
given intOrd as Ordering[Int]:
  ...
given listOrd[T: Ordering] as Ordering[List[T]]:
  ...

given Ordering[Int]:
  ...
given [T: Ordering] as Ordering[List[T]]:
  ...

given global as ExecutionContext = ForkJoinContext()
given Context = ctx
```

Here is how the above is expressed now:

```scala
given intOrd: Ordering[Int] with
  ...
given listOrd[T: Ordering]: Ordering[List[T]] with
  ...

given Ordering[Int] with
  ...
given [T: Ordering]: Ordering[List[T]] with
  ...

given global: ExecutionContext = ForkJoinContext()
given Context = ctx
```

> You can find an extensive discussion on the above change in its [PR #10538](https://github.com/lampepfl/dotty/pull/10538).

Reviewer 2: ```suggestion
You can find a discussion of the above change in the [PR #10538](https://github.com/lampepfl/dotty/pull/10538).
```

## Drop `as` in patterns

> Since we dropped `as` from `given`s, we lost a strong reason for having `as` at all. For this reason, we dropped them also from patterns. If previously the correct binding syntax is patterns was:

Reviewer 2: ```suggestion
Since we dropped `as` from `given`s, we lost a strong reason for having `as` at all. Therefore, we dropped `as` from patterns as well. The following syntax, valid in Scala 3.0.0-M2, is not accepted anymore:
```

```scala
case opt as Some(foo)
```

> Now it got reverted to the Scala 2 style:

Reviewer 2: ```suggestion
Instead, code that used it should be reverted to use the existing syntax of Scala 2, using `@`:
```

```scala
case opt @ Some(foo)
```

## Switch back to the old context function closure syntax
Previously, context function closures were written as follows:

```scala
(using s: Show[String]) => s.show("foobar")
```

Now, we have reverted to the following syntax:
```scala
(s: Show[String]) ?=> s.show("foobar")
```

# `Matchable` trait – a new top-level type

> The motivation for this change is the fact that it is possible to access the underlying type referenced by an opaque type using pattern matching:

Reviewer 2: ```suggestion
Although opaque types are supposed to provide an opaque abstraction of their underlying types, it is currently too easy to break the abstraction using pattern matching:
```

```scala
val imm = IArray(1,2,3) // supposedly immutable...
imm match

>    case a: Array[Int] => a(0) = 0  // but that's shown to be lie

Reviewer 2: ```suggestion
  case a: Array[Int] => a(0) = 0  // but that's shown to be lie
```

```

> To address this change, a new trait, `Matchable`, is introduced at the top of the type hierarchy:

Reviewer 2: ```suggestion
To address this change, we introduce a new trait `Matchable` near the top of the type hierarchy:
```

```scala
abstract class Any:
  def asInstanceOf
  def ==
  def !=
  def ##
  def equals
  def hashCode
  def toString

trait Matchable extends Any:
  def isInstanceOf
  def getClass

class AnyVal extends Any, Matchable

class Object extends Any, Matchable
```

> Matchable is currently a marker trait without any methods. Over time we might migrate methods getClass and isInstanceOf to it, since these are closely related to pattern-matching.

Reviewer 2: ```suggestion
`Matchable` is currently a marker trait without any methods. Over time, we intend to migrate the methods `getClass` and `isInstanceOf` to it, since these are closely related to pattern-matching.
```

> With the addition of this type, we now make the compiler emit warnings when trying to do a trick like the one discussed above:

Reviewer 2: ```suggestion
In the meantime, the compiler will emit warnings when trying to call those methods, or when performing a `match`, on an value that is not a `Matchable`. For example, the above example is now flagged:
```

> With the addition of this type, we now make the compiler emit warnings when trying to do a trick like the one discussed above:

Reviewer 8: this should mention that the warnings are only active with language mode `3.1-migration` or higher

```scala
-- Warning: i7314.scala:6:12 ---------------------------------------------------
6 |    case a: Array[Int] =>
  |            ^^^^^^^^^^
  |            pattern selector should be an instance of Matchable,
  |            but it has unmatchable type opaques.IArray[Int] instead
```

> The change involved an extensive discussion which you can read in the [PR #10670](https://github.com/lampepfl/dotty/pull/10670). Read more about this change in the [documentation](https://dotty.epfl.ch/docs/reference/other-new-features/matchable.html).

Reviewer 2: ```suggestion
You can read the discussion of this change in the [PR #10670](https://github.com/lampepfl/dotty/pull/10670). You can also read more about it in the [documentation](https://dotty.epfl.ch/docs/reference/other-new-features/matchable.html).
```

# Tooling improvements

> As the work on Scala 3 gets closer to the stable release, the focus increasingly shifts on the tooling available to get started with Scala 3.

Reviewer 2: ```suggestion
As we are getting closer to a stable release of Scala 3, the focus increasingly shifts on the tooling available to get started with Scala 3.
```

> For a while now, we are not using the old dottydoc documentation tool for building the documentation. We are developing an entirely new tool, scala3doc, from scratch.
> 
> As part of the tooling effort, this new Scala 3 documentation tool is rapidly improved. [PR #10522](https://github.com/lampepfl/dotty/pull/10522) proves that the doctool can generate documentation for the community build projects.

Reviewer 2: ```suggestion
As part of the tooling effort, we are rapidly improving the new Scala 3 documentation tool, `scala3doc`. As shown in the [PR #10522](https://github.com/lampepfl/dotty/pull/10522), it can now generate documentation for all the community build projects.
```

Author: We should stress here explicitly that the old tool is no longer in use.

> [PR #10491](https://github.com/lampepfl/dotty/pull/10491) introduced the scripting capability to Scala 3. Consider the following source named e.g. `Main.scala`:

Reviewer 2: ```suggestion
[PR #10491](https://github.com/lampepfl/dotty/pull/10491) introduced scripting support in Scala 3. Consider the following source named `Main.scala`:
```

```scala
@main def Test(name: String): Unit =
  println(s"Hello ${name}!")
```

> If you have Scala 3 binaries on your path (which you can get by following the steps at the [Dotty website](https://dotty.epfl.ch/)'s section "Try Dotty"), you can run the `scala Main.scala World` command. This command will compile the source in question to a temporary directory and run the discovered main method with the argument `World`.

Reviewer 2: ```suggestion
If you have Scala 3 binaries on your path (which you can get by following the steps on the [Dotty website](https://dotty.epfl.ch/), in the section "Try Dotty"), you can run the following command:

    $ scala Main.scala World

This will compile the source in question to a temporary directory and run the discovered main method with the argument `World`.
```

Note the difference from the Scala 2 scripting implementation. In Scala 2, we do not require the users to have a `main` method in their scripts due to it being too cumbersome to write. In Scala 3, thanks to the top-level definitions and the `@main` annotations, `main` methods are one-liners and hence are more suited for scripts.

The documentation for this feature is available [here](https://dotty.epfl.ch/docs/usage/getting-started.html#scala-3-for-scripting).

# Metaprogramming changes

> Metaprogramming-wise, the majority of work is focused on polishing the API and making it more uniform. The following notable metaprogramming API changes occurred between M2 and M3:

Reviewer 2: ```suggestion
We have been polishing the metaprogramming API and making it more uniform. The following notable changes occurred between M2 and M3:
```

- Fix #10359: Add GivenSelector to reflection API [#10469](https://github.com/lampepfl/dotty/pull/10469)
- Rework reflect show API [#10661](https://github.com/lampepfl/dotty/pull/10661)
- Fix #10709: Add missing level check before inlining [#10781](https://github.com/lampepfl/dotty/pull/10781)

# Let us know what you think!
If you have questions or any sort of feedback, feel free to send us a message on our
[Gitter channel](https://gitter.im/lampepfl/dotty). If you encounter a bug, please
[open an issue on GitHub](https://github.com/lampepfl/dotty/issues/new).

## Contributors
Thank you to all the contributors who made this release possible 🎉

According to `git shortlog -sn --no-merges 3.0.0-M2..3.0.0-M3` these are:

```
TODO
```

If you want to get your hands dirty and contribute to Dotty, now is a good time to get involved!
Head to our [Getting Started page for new contributors](https://dotty.epfl.ch/docs/contributing/getting-started.html),
and have a look at some of the [good first issues](https://github.com/lampepfl/dotty/issues?q=is%3Aissue+is%3Aopen+label%3Aexp%3Anovice).
They make perfect entry points into hacking on the compiler.

> We are looking forward to having you join the team of contributors.

Reviewer 2: Is this paragraph still relevant at this point?

Author: It's always good to have new passionate people who are ready to contribute. I have reworded this paragraph to reflect the current situation.

## Library authors: Join our community build

> Dotty now has a set of widely-used community libraries that are built against every nightly Dotty

Reviewer 2: ```suggestion
Scala 3 has a set of community libraries that are built against every nightly
```

> snapshot. Currently, this includes shapeless, ScalaPB, algebra, scalatest, scopt and squants.

Reviewer 2: This is obsolete.

Author: Reworded

```scala
-- Warning: i7314.scala:6:12 ---------------------------------------------------
6 |    case a: Array[Int] =>
  |            ^^^^^^^^^^
  |            pattern selector should be an instance of Matchable,
  |            but it has unmatchable type opaques.IArray[Int] instead
```

Note that the warnings are only active with language mode `3.1-migration` or higher - see the documentation on the [Language Versions](https://dotty.epfl.ch/docs/usage/language-versions.html) to learn how to enable it.

You can read the discussion of this change in the [PR #10670](https://github.com/lampepfl/dotty/pull/10670). You can also read more about it in the [documentation](https://dotty.epfl.ch/docs/reference/other-new-features/matchable.html).

# Tooling improvements
As we are getting closer to a stable release of Scala 3, the focus increasingly shifts on the tooling available to get started with Scala 3.

For a while now, we are not using the old dottydoc documentation tool for building the documentation. We are developing an entirely new tool, scala3doc, from scratch.

> As part of the tooling effort, this new Scala 3 documentation tool is rapidly improved. [PR #10522](https://github.com/lampepfl/dotty/pull/10522) proves that the doctool can generate documentation for the community build projects.

Reviewer 4: We can link here to a summaty page of all generated docs from community build: https://scala3doc.virtuslab.com/pr-master-docs/index.html

Reviewer 4: We can also mention that scala3doc is not as slow and fragile as dottydos.
